### DISCUSSION TITLE
[BUG] Connect-PnPOnline throws "Could not load file or assembly 'Microsoft..Identity.Client'

### DISCUSSION THREAD

**Discussion by SpodoY** [pnp/powershell#3201] [General]
### Reporting an Issue or Missing Feature

When I try to execute 'Connect-PnPOnline' with a given URL and a PSCredential Object I get the following error:

### Expected behavior 
I get a connection to my site

### Actual behavior
![image](https://github.com/pnp/powershell/assets/72346498/86d3e924-d026-401a-bfe9-07600611759b)

### Steps to reproduce behavior
`$credObject = New-Object System.Management.Automation.PSCredential -ArgumentList $username, $securePwd`
`Connect-PnPOnline -Url $SiteURL -Credentials $credObject` 

### What is the version of the Cmdlet module you are running?
![image](https://github.com/pnp/powershell/assets/72346498/e06df097-b2c0-49e4-ba43-0bb11af4d86e)

![image](https://github.com/pnp/powershell/assets/72346498/34fa172a-fc96-4f6d-b171-5be1975adb77)

### Which operating system/environment are you running PnP PowerShell on?
- [x] Windows

**Comment by gautamdsheth** [comment]
hi @SpodoY , looks like you might be using PnP PowerShell in combination with some other modules such as Az or Graph.

Can you please try using it standalone and then try , i.e just open PowerShell 7.2.x or higher and then enter the `Connect-PnPOnline` cmdlet ?

**Reply by SpodoY** [reply]
Hi @gautamdsheth , I just tested it standalone and it seems to work fine. The only other task I do inside my script is connect to a SQL Server ab execute SQL Commands via 'Invoke-Sqlcmd', but I don't see how this affects the Sharepoint connection to the extent where it just doesn't work anymore.

**Reply by santal-fuego** [reply]
I had the same error, and found that it is help to make Import-Module PnP... -Force
If there are other Import-Module lines, then the one with Import PnP should be the first one.

**Reply by SpodoY** [reply]
Thanks for the reply! Sadly this didn't fix my issue :(
I tried to put it as the first import and added -force but I still get the error

**Reply by AndersRask** [reply]
Using -force does nothing for this issue, since it only unloads and loads a module, but does not unload assemblies from appdomain, which is the reason for all the issues between different modules.

You can test this by loading assemblies from appdomain, importing module, removing the module and again load assemblies:

```
[System.AppDomain]::CurrentDomain.GetAssemblies()
ipmo pnp.powershell
rmo pnp.powershell
[System.AppDomain]::CurrentDomain.GetAssemblies()
```

**Comment by SpodoY** [comment]
As a follow up, it seems like the the PnP.Powershell Module has problems when used in conjunction with the SqlServer Module. When it split my script in two seperate scripts (one for the database query, one for PnP) it suddenly worked.

@gautamdsheth Can I provide any additional info so you could verify this?

**Comment by TomCaat** [comment]
I have encountered the same issue. The following is the detail description. 

When I use the Microsoft.Online.SharePoint.PowerShell and SharePointPnPPowerShellOnline modules together and try to import them, the PnP module always failed to be loaded. 
The root reason is that 
Microsoft.Online.SharePoint.PowerShell has a dependency on an internal module named as Microsoft.Online.SharePoint.Client.Tenant for version 16.0.0.0
SharePointPnPPowerShellOnline has a dependency on an internal module named as Microsoft.Online.SharePoint.Client.Tenant for version 16.1.0.0. 

This same name internal module of Microsoft.Online.SharePoint.Client.Tenant with different versions may bring out the failure of loading. 

Please take a check. I hope this will be a little helpful for you. @gautamdsheth

**Reply by AndersRask** [reply]
Please note, that step 1 in upgrading from SharePointPnPPowerShellOnline to PnP.PowerShell is, that you need to uninstall the old module first.

https://pnp.github.io/powershell/articles/upgrading.html?q=SharePointPnPPowerShellOnline
![image](https://github.com/pnp/powershell/assets/1651150/ce96e4ef-3b84-431e-9939-e2b6c784096c)

**Reply by TomCaat** [reply]
Dear AndersRask, 

The PnP.Powershell module is able to be loaded but the Connect-SPOservice is not able to run. It has the following error message. 
I am doubting that if this is a new conflict dependency module. 
_Could not load type 'Microsoft.SharePoint.Client.SharePointOnlineCredentials' from assembly 'Microsoft.SharePoint.Client.Runtime, Version=16.1.0.0, Culture=neutral, PublicKeyToken=71e9bce111e9429c'._

**My script:**
$PSVersionTable.PSVersion
Uninstall-Module -Name "SharePointPnPPowerShellOnline" -AllVersions -Force
Import-Module -Name "Microsoft.Online.SharePoint.PowerShell"
Import-Module -Name "PnP.PowerShell"

$siteUrl = 'https://TEST.sharepoint.com'
Connect-PnPOnline -ManagedIdentity -Url $siteUrl
#SharePointPnPPowerShellOnline 

get-pnpContext
#SharePointPnPPowerShellOnline

Connect-SPOservice -Url $siteUrl
#Microsoft.Online.SharePoint.PowerShell
Get-SPOHomeSite
#Microsoft.Online.SharePoint.PowerShell

**The issue:**
Connect-SPOService: untitled:Untitled-1:15:1
Line |
  15 |  Connect-SPOservice -Url $siteUrl
     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Could not load type 'Microsoft.SharePoint.Client.SharePointOnlineCredentials' from assembly 'Microsoft.SharePoint.Client.Runtime, Version=16.1.0.0, Culture=neutral, PublicKeyToken=71e9bce111e9429c'.

It seems that the PnP module is still not able to be imported with the Microsoft.Online.SharePoint.PowerShell module. 
Please kindly help take a look. Thanks a lot.

**Reply by TomCaat** [reply]
By the way, I have tried to import this PnP.PowerShell module to Azure Automation and encountered some issues as well.
The issue seems to be related with the module content as well.
(I have uninstalled the old modules before installing this.)

Error message:
Error importing the module PnP.PowerShell. Import failed with the following error: Orchestrator.Shared.AsyncModuleImport.ModuleImportException: While importing the module, an error occurred while processing the module content. Internal error message: Could not load file or assembly 'System.Management.Automation, Version=7.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.. at Orchestrator.Activities.SetModuleVersion.ExecuteInternal(CodeActivityContext context, String moduleName, ModuleLanguage moduleLanguage, String pythonModuleVersion, String modulePath, Guid accountId, Guid moduleVersionId, Int64 moduleContentByteSize, String storageUri, Int32 moduleVersion) at Orchestrator.Activities.SetModuleVersion.Execute(CodeActivityContext context) at System.Activities.CodeActivity.InternalExecute(ActivityInstance instance, ActivityExecutor executor, BookmarkManager bookmarkManager) at System.Activities.Runtime.ActivityExecutor.ExecuteActivityWorkItem.ExecuteBody(ActivityExecutor executor, BookmarkManager bookmarkManager, Location resultLocation)

I hope this will be helpful. I really expect for your assistance and ideas. Thanks a lot lot.

**Reply by AndersRask** [reply]
Yes, if you search the discussions, there are a lot of similar issues, so this is a well known issue.
It is a common PowerShell issue, when modules share assemblies of slightly different versions, and not really something there is a lot to do about, as the module is constructed (PowerShell 7 and core gives some possibilities for assemblies to co-exist, but that would require pnp.powershell module to be reworked completely, which is a lot to ask of the developers as they do this on their own time).

**Reply by AndersRask** [reply]
There really is no good solution other than keeping sessions seperated: one runspace to do your pnp stuff, and another for Az (the Start-ThreadJob/Start-Job suggestions in the documentation does not work!)

**Reply by TomCaat** [reply]
Thanks for your kind reply and suggestion, Rask. I really appreciate you and your project colleague's efforts. 
I can understand this issue that it is too hard to be repaired in a short term. Thanks again.

**Comment by TomCaat** [comment]
Greatly appreciate your comments.
I will try the new module again. Thanks again.

**Comment by AndersRask** [comment]
This issue should now have been fixed in latest nightly build:

https://github.com/pnp/powershell/issues/3207

**Comment by TomCaat** [comment]
这是来自QQ邮箱的自动回复邮件。
 
您好，您的邮件我已经收到！十分感谢来信！

**Comment by AndersRask** [comment]
The issue with colliding assemblies should now be fixed!
https://github.com/pnp/powershell/issues/3207
